scheduler/system_util.go

```diff
@@ -288,7 +309,78 @@ func diffSystemAllocsForNode(
 			result.place = append(result.place, allocTuple)
 		}
 	}
-	return result
+
+	finalResult := new(diffResult)
+	for _, result := range results {
+		finalResult.Append(result)
+	}
+
+	return finalResult
+}
+
+// ensureMaxSystemAllocCount enforces the invariant that the per-node diffResult
+// we have for a system or sysbatch job should never have more than "count"
+// allocations in a desired-running state.
+func ensureMaxSystemAllocCount(result *diffResult, count int) {
+
+	// sort descending by JobModifyIndex, then CreateIndex. The inputs are the
+	// ignore/update/reconnecting allocations for a single task group that were
+	// assigned to a single node, so that constrains the size of these slices
+	// and sorting won't be too expensive
+	sortTuples := func(tuples []allocTuple) {
+		sort.Slice(tuples, func(i, j int) bool {
+			I, J := tuples[i].Alloc, tuples[j].Alloc
+			if I.Job.JobModifyIndex == J.Job.JobModifyIndex {
+				return I.CreateIndex > J.CreateIndex
+			}
+			return I.Job.JobModifyIndex > J.Job.JobModifyIndex
+		})
+	}
+
+	// ignored allocs are current, so pick the most recent one and stop all the
+	// rest of the running allocs on the node
+	if len(result.ignore) > 0 {
+		if len(result.ignore) > count {
+			sortTuples(result.ignore)
+			result.stop = append(result.stop, result.ignore[count:]...)
+			result.ignore = result.ignore[:count]
+		}
+		count = count - len(result.ignore) // reduce the remaining count
+		if count < 1 {
+			result.stop = append(result.stop, result.update...)
+			result.stop = append(result.stop, result.reconnecting...)
+			result.update = []allocTuple{}
+			result.reconnecting = []allocTuple{}
+			return
+		}
+	}
+
+	// updated allocs are for updates of the job (in-place or destructive), so
+	// we can pick the most recent one and stop all the rest of the running
+	// allocs on the node
+	if len(result.update) > 0 {
```

Comment:
Something has been bugging me about this PR for the last few days and I finally realized what it was: we determine destructive vs non-destructive updates _after_ this per-node call is made and then limit the allocation changes based on `MaxParallel`. 

If we treat the non-destructive and destructive updates the same here, I'm fairly certain we'll break the system not-really-deployments feature of `MaxParallel/Stagger`, which makes decisions only based on the destructive endpoints. For example, suppose we have 3 nodes with `MaxParallel = 1`. One of the nodes has both a destructive and non-destructive update to make, but the alloc getting the non-destructive update is newer. In a single pass, we could end up stopping the alloc with the destructive update, update the alloc with the non-destructive update, and update an alloc on another node with a destructive update. That'd be more than `MaxParallel = 1` destructive updates (inasmuch as a "stop" is destructive).

I think we need to move the destructive vs non-destructive checking into the diff code and have a separate diff field for destructive vs non-destructive updates. But that'll ripple out to the generic scheduler as well. I'm going to need a little bit of time to get this locked down, so I think it's going to have to miss 1.5.0 GA.